operator/pkg/ciliumendpointslice/manager.go

```diff
@@ -23,75 +18,31 @@ var (
-	cesInsertedAt time.Time
-}
-
 // operations is an interface to all operations that a CES manager can perform.
 type operations interface {
 	// External APIs to Insert/Remove CEP in local dataStore
-	InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint, ns string) string
-	updateCEPToCESMapping(cepName string, cesName string)
-	RemoveCEPFromCache(cepName string, baseDelay time.Duration)
-	removeCEPFromCES(cepName string, cesName string, baseDelay time.Duration, identity int64, checkIdentity bool)
-	// Supporting APIs to Insert/Remove CEP in local dataStore and effectively
-	// manages CES's.
-	getCESFromCache(cesName string) (*cilium_v2.CiliumEndpointSlice, error)
-	getCESCopyFromCache(cesName string) (*cilium_v2.CiliumEndpointSlice, error)
-	updateCESInCache(ces *cilium_v2.CiliumEndpointSlice, deepCopy bool)
-	deleteCESFromCache(cesName string)
-	getRemovedCEPs(string) map[string]struct{}
-	clearRemovedCEPs(string, map[string]struct{})
-	createCES(cesName string) *cesTracker
-	addCEPtoCES(cep *cilium_v2.CoreCiliumEndpoint, ces *cesTracker)
-	insertCESInWorkQueue(ces *cesTracker, baseDelay time.Duration)
-	// APIs to collect metrics of CES and CEP
-	getTotalCEPCount() int
-	getCEPCountInCES(cesName string) int
-	getCESCount() int
-	getAllCESs() []cesOperations
-	getAllCEPNames() []string
-	getCESMetricCountersAndClear(cesName string) (cepInsert int64, cepRemove int64)
-	getCESQueueDelayInSeconds(cesName string) (diff float64)
-}
+	UpdateCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns string) []CESName
+	RemoveCEPMapping(cep *cilium_v2.CoreCiliumEndpoint, ns string) CESName
+
+	initializeMappingForCES(ces *cilium_v2.CiliumEndpointSlice) CESName
+	initializeMappingCEPtoCES(cep *cilium_v2.CoreCiliumEndpoint, ns string, ces CESName)
 
-type cesOperations interface {
-	getAllCEPs() []cilium_v2.CoreCiliumEndpoint
-	getCESName() string
-	getCEPNameFromCCEP(cep *cilium_v2.CoreCiliumEndpoint) string
+	getCEPCountInCES(ces CESName) int
+	getCEPinCES(ces CESName) []CEPName
+	getCESData(ces CESName) CESData
+	isCEPinCES(cep CEPName, ces CESName) bool
 }
 
 // cesMgr is used to batch CEP into a CES, based on FirstComeFirstServe. If a new CEP
 // is inserted, then the CEP is queued in any one of the available CES. CEPs are
 // inserted into CESs without any preference or any priority.
 type cesMgr struct {
-
-	// desiredCESs is used to map CESName to CESTracker[i.e. list of CEPs],
+	logger logrus.FieldLogger
+	// mapping is used to map CESName to CESTracker[i.e. list of CEPs],
 	// as well as CEPName to CESName.
-	desiredCESs *CESToCEPMapping
+	mapping *CESToCEPMapping
 
-	// workqueue is used to sync CESs with the api-server. this will rate-limit the
-	// CES requests going to api-server, ensures a single CES will not be proccessed
-	// multiple times concurrently, and if CES is added multiple times before it
-	// can be processed, this will only be processed only once.
-	queue workqueue.RateLimitingInterface
+	mutex lock.Mutex
```

Comment:
This lock can be avoided since `mapping` is assigned once at the manager startup and then every ops on it is protected by its internal lock. Same for `maxCEPsInCES`, it is assigned at startup and then only read.

Reply:
It is used in the following cases:

Get largest not full CES and create new one if there is none. Without lock there may be too many CEPs in CES or too many CESs (all gets the same CES and fill it above capacity or all create new one when one would be enough)
```go
cesName = c.getLargestAvailableCESForNamespace(ns)
if cesName.Name == "" {
	cesName = c.createCES("", ns)
}
c.mapping.insertCEP(cepName, cesName)
```

Check if CES can be deleted. Without this check new CEP may be added to the CES and CES then deleted.
```go
c.mutex.Lock()
defer c.mutex.Unlock()
if c.mapping.countCEPsInCES(cesName) == 0 {
	c.mapping.deleteCES(cesName)
}
```

The same, a little bit more complicated, for identity.

I think I could rewrite this code without this lock. The mapping could do the checks and then return error. But it means the code would need to be changed to the loop.

Reply:
> It is used in the following cases:
> 
> Get largest not full CES and create new one if there is none. Without lock there may be too many CEPs in CES or too many CESs (all gets the same CES and fill it above capacity or all create new one when one would be enough)
> 

But the only operations that can insert or remove a CEP are:
- `UpdateCEPMapping` and `RemoveCEPMapping`, but these are serialized by the channel returned by `Events` in `runCiliumEndpointsUpdater` (IOW it's not possible to handle multiple CiliumEndpoint related events at the same time)
- `initializeMappingCEPtoCES` that is called only before initial sync is completed

So, it should not be possible to have multiple goroutines concurrently reading the CEPs in CES counter and trying to  insert or remove a CEP in a CES. Or am I missing something?

Reply:
Yes. As long as we are sure we cannot process multiple events from the CEP at the same time we are good here. I'm removing those locks for now.

I thought that the CES controller worker also changes mapping but it's not the case, it's:

CEP event => update mapping => enqueue CES => worker processing CESs from the queue
CES event => enqueue CES => the same worker as above processing CESs from the queu

I'll remove the locks.

Reply:
Removed